### DISCUSSION TITLE
getStaticProps thoughts

### DISCUSSION THREAD

**Discussion by samuelcastro** [UnlyEd/next-right-now#359] [General]
I'm trying to take advantage of `getStaticProps` to execute data manipulation logics and return data exactly how it's going to be consume by the client, and even though this might have some performance benefits (server side vs client side) I think it's keeping the code hard to maintain and reason about, and the performance benefits might not really be significant.

I'm curious to hear you feedback on this. What do you usually do?

Say you have a big array of objects you fetched on `getStaticProps` and you need to parse it in a specific format to be used in a `react-select` kind of component in the client, where would you do this? (getStaticProps or client)

**Comment by Vadorequest** [comment]
It really depends on the data processing.

I've had this challenge in my own app using the `-at` preset, where I fetch tons of data from the server (fetching 19 tables one by one because... REST-ish). After fetching, I process the data in many ways, sanitizing, resolving relationships, applying default values. This processing is "heavy" and I thought it'd be better to do it in the server.

At some point, I ran into the "circular dependency issue" with relationships. Because I was resolving those on the server and had to send them to the client (stringified), I had to break the circular issue. At first, I wrote a "limiter" in the relationships resolver, which would limit to resolving 3 levels of circular references. It wasn't such a great idea, as it broke my app randomly depending on how I was manipulating the data. I had to increment it to 4, then 5, and then I noticed how much KB were sent to the browser and it was insane and unpredictable.

Eventually I threw that away and rethought the whole processing.

At this time, I still perform data sanitization and defaults values on the server, but I resolve the relationships on the client. There, I benefit from the "circular reference **feature**" and I don't have to bother about knowing whether I manipulate the root objects or nested objects, it simply works.

**Reply by samuelcastro** [reply]
Yeah that make sense, thanks for sharing that. I'm just trying to balance the trade off of running everything in the server vs the maintainability benefits of having manipulation done where it's actually being used (single responsibility pattern).

But I think there is no right answer.

**Reply by Vadorequest** [reply]
I guess the right answer is the one that fits your need at a given time. 😉
